server/src/main/java/com/google/finapp/SpannerDaoImpl.java

```diff
@@ -251,20 +234,49 @@ public BigDecimal createTransactionForAccount(
     }
   }
 
-  private ImmutableMap<ByteArray, BigDecimal> readAccountBalances(
-      ByteArray fromAccountId, ByteArray toAccountId, TransactionContext transaction) {
+  /**
+   * Returns an ImmutableMap of all requested AccountData keyed by account ids.
+   *
+   * @param accountIds account ids for which to return AccountData
+   * @param transaction the transaction context to use
+   * @return mapping of AccountData, keyed by account id
+   * @throws StatusException if any of the requested accountIds is not found, or the associated
+   *     accountStatus is not ACTIVE.
+   */
+  private ImmutableMap<ByteArray, AccountData> readAccountDataForTransfer(
+      Iterable<ByteArray> accountIds, TransactionContext transaction) throws StatusException {
+    KeySet.Builder keySetBuilder = KeySet.newBuilder();
+    for (ByteArray accountId : accountIds) {
+      keySetBuilder.addKey(Key.of(accountId));
+    }
     ResultSet resultSet =
         transaction.read(
             "Account",
-            KeySet.newBuilder().addKey(Key.of(fromAccountId)).addKey(Key.of(toAccountId)).build(),
-            ImmutableList.of("AccountId", "Balance"));
+            keySetBuilder.build(),
+            ImmutableList.of("AccountId", "AccountStatus", "Balance"));
 
-    ImmutableMap.Builder<ByteArray, BigDecimal> accountBalancesBuilder = ImmutableMap.builder();
+    ImmutableMap.Builder<ByteArray, AccountData> accountDataBuilder = ImmutableMap.builder();
     while (resultSet.next()) {
-      accountBalancesBuilder.put(
-          resultSet.getBytes("AccountId"), resultSet.getBigDecimal("Balance"));
+      AccountData accountData = new AccountData();
+      accountData.balance = resultSet.getBigDecimal("Balance");
+      accountData.status = AccountStatus.forNumber((int) resultSet.getLong("AccountStatus"));
+      accountDataBuilder.put(resultSet.getBytes("AccountId"), accountData);
+    }
+    ImmutableMap<ByteArray, AccountData> accountData = accountDataBuilder.build();
+    for (ByteArray accountId : accountIds) {
+      if (!accountData.containsKey(accountId)) {
+        throw Status.INVALID_ARGUMENT
+            .withDescription(String.format("Account not found: %s", accountId.toString()))
+            .asException();
+      } else if (accountData.get(accountId).status != AccountStatus.ACTIVE) {
+        throw Status.INVALID_ARGUMENT
+            .withDescription(
+                String.format(
+                    "Non-active accounts are not eligible for transfers: %s", accountId.toString()))
+            .asException();
+      }
```

Comment:
This is a bit surprising since I'd assume `moveAccountBalance` is the part that cares whether the account is active.

Reply:
I do not understand the comment. `readAccountDataForTransfer` is used for both `moveAccountBalance` and `createTransactionForAccount`. Given this method returns account data `ForTransfer` it does make sense to move eligibility checks here no?

Reply:
I just mean that `readAccountDataForTransfer` could reasonably return data for non-active accounts based on the name, not a blocking comment.

Reply:
You are right, it could be either. Actually, if the method was `readAccountData` it *should not* do the check. I explicitly went with the `ForTransfer` naming so I can move validations inside the helper to avoid duplication.